rfcs/20201002-CTC-API-consistency.md

```diff
@@ -0,0 +1,227 @@
+# Change the CTC-related APIs to ensure the consistency of usage
+
+| Status        |  Proposed                                            |
+:-------------- |:---------------------------------------------------- |
+| **RFC #**     | [295](https://github.com/tensorflow/community/pull/295)|
+| **Author(s)** | Mingjie Zhou (benjamin_chou@outlook.com), |
+| **Sponsor**   | François Chollet (francois.chollet@gmail.com)                 |
+| **Updated**   | 2020-10-02                                           |
+
+## Objective
+Change the APIs of connectionist temporal classification (CTC) decoders (beam search decoder and greedy decoder) to ensure the consistency with the usage of CTC loss. 
+Preserve the same arguments of API as far as possible.
+
+
+## Motivation
+When implementing a model for automatic speech recognition (ASR) and optical character recognition (OCR), one may need to apply CTC at the last layer of their models. 
+However, the training behavior is not the same as that of testing.
+During training, one needs to apply the CTC loss as a part of the objective function.
+During testing, the beam search decoder or greedy decoder will be used to get the labels with the maximum probability. 
+
+The CTC introduces a **blank** label, which one needs to specify the index of it in their model, to separate the different characters. The current APIs of CTC loss and its decoders don't use the same blank index by default and the API documentation doesn't point it out, which may make developers and reseachers confused and lead to an invisible mistake.
+
+
+## User Benefit
+This, apparently, allows developers and researchers to use CTC API easily without making a mistake.
+
+## Design Proposal
+The expectation is to change the original APIs to the proposed APIs.
+
+[ctc_loss](https://www.tensorflow.org/api_docs/python/tf/nn/ctc_loss):
+<pre><code>tf.nn.ctc_loss(
+    labels, logits, label_length, logit_length, logits_time_major=True,unique=None, 
+    blank_index=None, name=None
+)</code></pre>
+According to API docs, default blank label is 0 rather num_classes - 1, unless overridden by blank_index.
+
+**Original APIs**:
+Both beam search decoder and greedy decoder perform with `blank index = num_classes - 1` and `logits_time_major=True` by default behavior without an option in the arguments. 
+However, the blank index is different from the default in `tf.nn.ctc_loss` and no reminder in the API doc.
```

Comment:
Hmm; looking at the original CTCLoss flow, it looks like `tf.nn.ctc_loss` uses `blank_index = num_classes - 1` [here](https://github.com/tensorflow/tensorflow/blob/c2c79377f39c64d0130e8209f1fb6454b166e787/tensorflow/core/kernels/ctc_loss_op.cc#L198).  The "CTCLossV2" op, which only kicks in if `use_cudnn` is passed, [uses the broken blank_index default](https://github.com/tensorflow/tensorflow/blob/3e63843a2b55c428ab2833b4054c5acfc9827184/tensorflow/python/ops/ctc_ops.py#L222).

In my opinion, this was a bug when the cudnn version of the loss was implemented.  I dug down to [this call](https://github.com/tensorflow/tensorflow/blob/b95f88a5e759367ae80bad988fdd35cbd0f604b4/tensorflow/stream_executor/cuda/cuda_dnn.cc#L2007); and looked through the cudnn [documentation](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnCTCLoss) on the associated functions.  It doesn't look like they provide a way to describe which index is the blank.

To fix this **bug**, the solution would be to have the cudnn version swap the provided `blank_index` with the 0th index in this case, if we cannot pass a proper blank index to the cudnn code.

Reply:
> Hmm; looking at the original CTCLoss flow, it looks like `tf.nn.ctc_loss` uses `blank_index = num_classes - 1` [here](https://github.com/tensorflow/tensorflow/blob/c2c79377f39c64d0130e8209f1fb6454b166e787/tensorflow/core/kernels/ctc_loss_op.cc#L198). The "CTCLossV2" op, which only kicks in if `use_cudnn` is passed, [uses the incorrect blank_index of 0](https://github.com/tensorflow/tensorflow/blob/3e63843a2b55c428ab2833b4054c5acfc9827184/tensorflow/python/ops/ctc_ops.py#L222).
> 
> In my opinion, this was a bug when the cudnn version of the loss was implemented as it introduce differing behavior if and only if the `use_cudnn` flag is provided. I dug down to [this call](https://github.com/tensorflow/tensorflow/blob/b95f88a5e759367ae80bad988fdd35cbd0f604b4/tensorflow/stream_executor/cuda/cuda_dnn.cc#L2007); and looked through the cudnn [documentation](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnCTCLoss) on the associated functions. It doesn't look like they provide a way to describe which index is the blank.
> 
> To fix this **bug**, the solution would be to have the cudnn version swap the provided `blank_index` with the 0th index in this case, if we cannot pass a proper blank index to the cudnn code.

From my perspective, this wasn't a bug. Actually, the blank index is shifted before passed to cudnn [here](https://github.com/tensorflow/tensorflow/blob/3e63843a2b55c428ab2833b4054c5acfc9827184/tensorflow/python/ops/ctc_ops.py#L715).
Looking at the original CTCLoss workflow:
- For sparse tensor, it will use `_ctc_loss_op_standard`(CPU) or `_ctc_loss_op_cudnn`(GPU). The `_ctc_loss_op_standard` by default uses the `num_classes - 1` as the blank index, while the `_ctc_loss_op_cudnn` uses `0` as the blank index. Both implementations have shifted the blank index to the corresponding position before passing arguments to `_ctc_loss_impl`. 

<pre><code>def _ctc_loss_op_standard(labels, logits, logit_length, logits_time_major,
                          blank_index):
part_before = logits[:, :, :blank_index]
part_after = logits[:, :, blank_index + 1:]
part_blank = logits[:, :, blank_index:blank_index + 1]
logits = array_ops.concat([part_before, part_after, part_blank], axis=2)
...
</code></pre>

<pre><code>def _ctc_loss_op_cudnn(labels, logits, logit_length, logits_time_major,
                       blank_index):
  part_before = logits[:, :, :blank_index]
  part_after = logits[:, :, blank_index + 1:]
  part_blank = logits[:, :, blank_index:blank_index + 1]
  logits = array_ops.concat([part_blank, part_before, part_after], axis=2)
...
</code></pre>
- For dense tensor, the [dense version](https://github.com/tensorflow/tensorflow/blob/3e63843a2b55c428ab2833b4054c5acfc9827184/tensorflow/python/ops/ctc_ops.py#L939) is implemented by python code with custom gradients.  This dense implementation performs with `blank_index=0` by default. For `blank_index!=0` cases, it would be shifted at the beginning [here](https://github.com/tensorflow/tensorflow/blob/3e63843a2b55c428ab2833b4054c5acfc9827184/tensorflow/python/ops/ctc_ops.py#L1014). 

Here are my suggestions:
- Improve the `tf.nn.ctc_loss` documentation to make it clear that what kind of situation will use which blank index by default.  And provide some examples.
- Improve the `tf.nn.ctc_beam_search_decoder` and `tf.nn.ctc_greedy_decoder` to allow users to specify blank index as described in the RFC.

Reply:
Yes; I think your approach is right.  I'm sad that the API has bifurcated this way but at this point it's too late to take it back.